perf/src/recycler.rs

```diff
@@ -289,63 +88,60 @@ impl<T: Default + Reset + Sized> Recycler<T> {
+    pub fn allocate(&self, name: &'static str) -> T {
+        {
+            const RECYCLER_SHRINK_WINDOW_HALF: usize = RECYCLER_SHRINK_WINDOW / 2;
+            const RECYCLER_SHRINK_WINDOW_SUB_ONE: usize = RECYCLER_SHRINK_WINDOW - 1;
+            let mut gc = self.recycler.gc.lock().unwrap();
+            // Update the exponential moving average of gc.len().
+            self.recycler.size_factor.store(
+                self.recycler
+                    .size_factor
+                    .load(Ordering::Acquire)
+                    .saturating_mul(RECYCLER_SHRINK_WINDOW_SUB_ONE)
+                    .saturating_add(RECYCLER_SHRINK_WINDOW_HALF)
+                    .checked_div(RECYCLER_SHRINK_WINDOW)
+                    .unwrap()
+                    .saturating_add(gc.len()),
+                Ordering::Release,
+            );
+            if let Some(mut x) = gc.pop() {
                 self.recycler.stats.reuse.fetch_add(1, Ordering::Relaxed);
-                reused_object.reset();
-                Some(reused_object)
+                x.reset();
+                return x;
             }
-            AllocationDecision::Allocate(total_allocated_count, recycled_len) => {
-                let mut t = T::default();
-                t.set_recycler(Arc::downgrade(&self.recycler));
-                if total_allocated_count % 1000 == 0 {
-                    datapoint_info!(
-                        "recycler_total_allocated_count",
-                        ("name", self.shrink_metric_name, String),
-                        ("count", total_allocated_count as i64, i64),
-                        ("recycled_len", recycled_len as i64, i64),
-                    )
-                }
-                Some(t)
-            }
-
-            AllocationDecision::AllocationLimitReached => None,
         }
-    }
+        let total = self.recycler.stats.total.fetch_add(1, Ordering::Relaxed);
+        trace!(
+            "allocating new: total {} {:?} id: {} reuse: {} max_gc: {}",
+            total,
+            name,
+            self.recycler.id,
+            self.recycler.stats.reuse.load(Ordering::Relaxed),
+            self.recycler.stats.max_gc.load(Ordering::Relaxed),
+        );
 
-    pub fn recycle_for_test(&self, x: T) {
-        self.recycler.recycle(x);
+        let mut t = T::default();
+        t.set_recycler(Arc::downgrade(&self.recycler));
+        t
     }
 }
 
 impl<T: Default + Reset> RecyclerX<T> {
     pub fn recycle(&self, x: T) {
         let len = {
             let mut gc = self.gc.lock().expect("recycler lock in pub fn recycle");
-            gc.objects.push(x);
+            gc.push(x);
+            const SIZE_FACTOR_AFTER_SHRINK: usize = RECYCLER_SHRINK_SIZE * RECYCLER_SHRINK_WINDOW;
+            if gc.len() > RECYCLER_SHRINK_SIZE
+                && self.size_factor.load(Ordering::Acquire) >= SIZE_FACTOR_AFTER_SHRINK
+            {
+                for mut x in gc.drain(RECYCLER_SHRINK_SIZE..) {
```

Comment:
Does this mean once the recycler goes up to say `n * RECYCLER_SHRINK_SIZE`, then it will either 

1) Stay at `n * RECYCLER_SHRINK_SIZE` or
2) Shrink to `RECYCLER_SHRINK_SIZE`

I can imagine a case where we end up only using some `m * RECYCLER_SHRINK_ SIZE` amount of memory and `1 < m < n`, where it might be advantageous to shrink, but not shrink *all* the way down.

Given a `size_factor`, could we shrink down, for instance, to the closest multiple of `RECYCLER_SHRINK_SIZE`?